Add SQUASHFS_MOUNT_OFFSET enviroment varialbe

dleggo opened the pull request:
This would allow the offset value to be passed in through an enviroment variable `SQUASHFS_MOUNT_OFFSET`. The call to `strtoul` should sanatize any potentially malicious values being passed in. This feature is quite useful when there is other data (or code) before the squashfs file that needs to be ignored.

dleggo committed:
Patch to allow offset to be passed in through enviroment variable

SQFSMNT_OFFSET

dleggo committed:
Patch to allow offset to be passed in through enviroment variable

SQFSMNT_OFFSET

dleggo committed:
Merge

dleggo committed:
Rename enviroment variable

bcumming reviewed (changes requested):
There are a couple of issues with this approach of using an environment variable:

1. `squashfs-mount` can mount multiple squashfs images at different mount points. The environment variable is a single value: for full flexibility, we must be able to specify a unique offset for each squashfs file.
2. If the variable is set, the behavior of the squashfs-mount executable changes - this is not acceptable for our use case.

Given this, the offset could be provided as part of the command line argument (optionally) and clearly associated with a specific squashfs file.

The current syntax provides the mount points as positional arguments, with a `:` to separate the squashfs file and the mount point:
```
squashfs-mount $HOME/img1.sqfs:/tools $HOME/img1.sqfs:$HOME/python -- bash
```

This could be extended to allow an optional `@` that would be an offset in bytes:
```
squashfs-mount $HOME/img1.sqfs:/tools $HOME/img1.sqfs@2048:$HOME/python -- bash
```
where we have an offset of 2048 bytes for the second squashfs image.

* the offset would be a member of `mount_entry_t`
* it would be parsed in `parse_mount_entries`.
* it should be validated to be a non-negative integer that is less than the size of the squashfs image

dleggo committed:
Add offset as a per-mount argument

dleggo committed:
Add case for negative offset

dleggo commented:
I added the corrections you requested. Could you have a look? :smiley:

dleggo commented:
Sorry to bump, but could this get a look? 😃

simonpintarelli commented:
Hi @dleggo , sorry for the delay. I'll try to have a look this week.

dleggo commented:
No worries! Thank you!!

simonpintarelli committed:
fix parse offset logic

simonpintarelli committed:
stat returns 0 on success

simonpintarelli committed:
add a test

simonpintarelli committed:
update ubuntu

simonpintarelli committed:
update up/download-artifact@v3 -> 4

simonpintarelli committed:
update test and fix error checking

simonpintarelli commented:
Sorry for the delay @dleggo. The proposed format `<image>@<offset>:<mountpoint>` didn't work for me, I've changed the `parse_mount_entries` accordingly.

We are considering to rewrite this tool in C++, I'm not very confident with increasing complexity of input parsing in a setuid binary.

dleggo commented:
Sorry for the late reply.

> The proposed format <image>@<offset>:<mountpoint> didn't work for me, I've changed the parse_mount_entries accordingly.

Yeah, sorry. I added the offset option at the end instead of the middle. Thank you!

> We are considering to rewrite this tool in C++, I'm not very confident with increasing complexity of input parsing in a setuid binary.

Okay!
